### DISCUSSION TITLE
In SIMD context, how to use correctly conditional expressions with loops in C++?

### DISCUSSION THREAD

**Discussion by g1r4ff3** [mitsuba-renderer/mitsuba3#891] [Q&A] [answered]
Hi, I'm struggling to make a torus shape plugin for mitsuba 3 but while i'm trying to write loops such as _for_ in my code, i keep failing compilation due to conditional expressions within loop statements. below is the example code:

```
template <typename FloatP, typename Ray3fP>
std::tuple<FloatP, Point<FloatP, 2>, dr::uint32_array_t<FloatP>,
			 dr::uint32_array_t<FloatP>>
ray_intersect_preliminary_impl(const Ray3fP &ray_, dr::mask_t<FloatP> active) const {

...
some setup codes
...

// solve the quartic equation.
const Value NR_ROOTS = SolveQuartic(coeffs, roots);

// find a nearest root.
FloatP t                       = dr::Largest<FloatP>;
for (ScalarValue i = 0; active = active && dr::any(i < NR_ROOTS); ++i)
{
	if (active = active && dr::any(Value(0.0) <= roots[i] && roots[i] <= maxt))
	{
              t = dr::select(active, dr::select(roots[i] < t, roots[i], t), dr::Infinity<FloatP>);
	}
}

```

My intention of the code is finding a nearest root from the roots obtained by SolveQuartic().
Could you help me to correct above code please?

Thank you in advance.

**Comment by merlinND** [comment, marked as answer]
Hello @g1r4ff3,

If you have a small number of roots to iterate over, a standard C++ for loop is fine (it will lead be fully unrolled in the generated kernel, but that's okay for a small iteration count and loop body size).

I would then write something like:

```c++
FloatP t = dr::Largest<FloatP>;
for (size_t root_i = 0; root_i < n_roots; ++root_i) {  // <-- only scalar quantities here
    // Update entries of `t` for this this is the closest root
    Mask is_closest_root = ...;  // your logic
    t = dr::select(is_closest_root, roots[i], t);
}
```

**Reply by g1r4ff3** [reply]
I wrote the code as you wrote and the code is compiled well. thank you!
